### PR TITLE
Adding Stackdriver Backend

### PR DESCRIPTION
**Pull request by winfieldj**
This PR adds a Stackdriver metrics backend. Collected Buildkite metrics are written to Stackdriver Custom Metrics at runtime.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -28,12 +28,13 @@ func main() {
...
 
 		// backend config
-		backendOpt     = flag.String("backend", "cloudwatch", "Specify the backend to use: cloudwatch, statsd, prometheus")
+		backendOpt     = flag.String("backend", "cloudwatch", "Specify the backend to use: cloudwatch, statsd, prometheus, stackdriver")
 		statsdHost     = flag.String("statsd-host", "127.0.0.1:8125", "Specify the StatsD server")
 		statsdTags     = flag.Bool("statsd-tags", false, "Whether your StatsD server supports tagging like Datadog")
 		prometheusAddr = flag.String("prometheus-addr", ":8080", "Prometheus metrics transport bind address")
 		prometheusPath = flag.String("prometheus-path", "/metrics", "Prometheus metrics transport path")
 		clwDimensions  = flag.String("cloudwatch-dimensions", "", "Cloudwatch dimensions to index metrics under, in the form of Key=Value, Other=Value")
+		projectID      = flag.String("stackdriver-projectID", "", "Specify Stackdriver projectID")
```
Comment (elestril) [inline main.go]: gcpProjectID ?

File: go.mod [hunk 2]
```diff
@@ -1,21 +1,15 @@
...
-	github.com/pmezard/go-difflib v1.0.0 // indirect
-	github.com/prometheus/client_golang v0.9.0
-	github.com/prometheus/client_model v0.0.0-20180712105110-5c3871d89910
-	github.com/prometheus/common v0.0.0-20181020173914-7e9e6cabbd39 // indirect
-	github.com/prometheus/procfs v0.0.0-20181005140218-185b4288413d // indirect
-	github.com/stretchr/testify v1.2.2 // indirect
-	golang.org/x/net v0.0.0-20181029044818-c44066c5c816 // indirect
-	golang.org/x/sync v0.0.0-20180314180146-1d60e4601c6f // indirect
-	golang.org/x/text v0.3.0 // indirect
+	github.com/prometheus/client_golang v0.9.3-0.20190127221311-3c4408c8b829
```
Comment (elestril) [inline go.mod:10]: those version bumps look unrelated. You may want to revert them to keep the PR small ?

File: backend/stackdriver.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	ctx := context.Background()
+	req := &monitoringpb.CreateTimeSeriesRequest{
+		Name: "projects/" + *projectID,
+		TimeSeries: []*monitoringpb.TimeSeries{{
+			Metric: &metricpb.Metric{
+				Type: *metricType,
+			},
+			Points: []*monitoringpb.Point{{
+				Interval: &monitoringpb.TimeInterval{
+					StartTime: time,
```
Comment (eberle1080) [inline backend/stackdriver.go:120]: I feel like StartTime having the same value as EndTime will result in some kind of divide-by-zero problem somewhere. Is there some documentation to explain this?

File: main.go (hunk 1 above)
Comment (winfieldj) [reply main.go]: Sure

File: go.mod (hunk 2 above)
Comment (winfieldj) [reply go.mod:10]: Good catch

File: go.sum [hunk 4]
```diff
@@ -1,7 +1,21 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (eberle1080) [inline go.sum:1]: Minor nit, you might want to run "go mod tidy" to remove unused entries

**Review by eberle1080 — APPROVED** [review]
Looks good. A few nits, but no blockers

File: README.md [hunk 5]
```diff
@@ -72,6 +72,10 @@ The Prometheus backend supports the following arguments
 * `-prometheus-addr`: The local address to listen on (defaults to `:8080`).
 * `-prometheus-path`: The path under `prometheus-addr` to expose metrics on (defaults to `/metrics`).
 
+The Stackdriver backend supports the following arguments
+
+* `-stackdriver-projectID`: The Google Cloud Platform to report metrics for
```
Comment (srmocher) [inline README.md]: As mentioned, `gcp-projectID` might be better.

File: main.go [hunk 6]
```diff
@@ -68,6 +69,13 @@ func main() {
 		}
 	case "prometheus":
 		bk = backend.NewPrometheusBackend(*prometheusPath, *prometheusAddr)
+	case "stackdriver":
+		var err error
+		bk, err = backend.NewStackdriverBackend(*projectID)
+		if err != nil {
+			fmt.Printf("Error starting Stackdriver, err: %v\n", err)
```
Comment (srmocher) [inline main.go]: Nit: `Error starting Stackdriver Backend` might be better?

**Comment by srmocher** [comment]
LGTM. Might be good to test and check in Stackdriver for the metrics before submitting this PR upstream.
